File: impl-csharp/src/Historical.cs
using System.Numerics;
using System.Text.Json.Serialization;

namespace NoaReceipt;

public sealed class HistoricalEvidence
{
    [JsonPropertyName("retirement")] public string Retirement { get; init; } = "NOT_PROVIDED";
    [JsonPropertyName("witness")] public string Witness { get; init; } = "NOT_PROVIDED";
    [JsonPropertyName("availability")] public string Availability { get; init; } = "NOT_PROVIDED";
}

public sealed class HistoricalDimensions
{
    [JsonPropertyName("integrity")] public string Integrity { get; init; } = "UNANSWERED";
    [JsonPropertyName("completeness")] public string Completeness { get; init; } = "UNANSWERED";
    [JsonPropertyName("attribution")] public string Attribution { get; init; } = "UNATTRIBUTABLE";
    [JsonPropertyName("organizationalIndependence")] public string OrganizationalIndependence { get; init; } = "UNVERIFIED";
    [JsonPropertyName("evidence")] public HistoricalEvidence Evidence { get; init; } = new();
}

public sealed class HistoricalPolicy
{
    [JsonPropertyName("verifierVersion")] public string VerifierVersion { get; init; } = HistoricalVerifier.Spec;
    [JsonPropertyName("purpose")] public string Purpose { get; init; } = "historical-audit";
}

public sealed class HistoricalResult
{
    [JsonPropertyName("spec")] public string Spec { get; init; } = HistoricalVerifier.Spec;
    [JsonPropertyName("policy")] public HistoricalPolicy Policy { get; init; } = new();
    [JsonPropertyName("classification")] public string Classification { get; init; } = "INVALID";
    [JsonPropertyName("code")] public string Code { get; init; } = "RECEIPT_MALFORMED";
    [JsonPropertyName("dimensions")] public HistoricalDimensions Dimensions { get; init; } = new();
    [JsonPropertyName("chain")] public string? Chain { get; init; }
    [JsonPropertyName("count")] public int Count { get; init; }
    [JsonPropertyName("attributedThroughSeq")] public long? AttributedThroughSeq { get; init; }
    [JsonPropertyName("asOf")] public string? AsOf { get; init; }
}

public static class HistoricalVerifier
{
    public const string Spec = "noa.historical-verification/0.1";
    private const string LifecycleSpec = "noa.signing-key-lifecycle/0.1";

    private sealed class HistoricalTrust
    {
        public required JObj Keyring { get; init; }
        public required Dictionary<string, bool> Retired { get; init; }
        public required Dictionary<string, BigInteger?> ValidFrom { get; init; }
        public required Dictionary<string, BigInteger?> RetiredAt { get; init; }
        public required bool Lifecycle { get; init; }
    }

    private static HistoricalDimensions Dimensions(
        string integrity,
        string completeness,
        string attribution,
        string retirement,
        string witness,
        string availability) => new()
    {
        Integrity = integrity,
        Completeness = completeness,
        Attribution = attribution,
        OrganizationalIndependence = "UNVERIFIED",
        Evidence = new HistoricalEvidence
        {
            Retirement = retirement,
            Witness = witness,
            Availability = availability,
        },
    };

    private static HistoricalResult Result(
        string classification,
        string code,
        HistoricalDimensions dimensions,
        string? chain,
        int count,
        long? attributedThroughSeq = null,
        string? asOf = null) => new()
    {
        Classification = classification,
        Code = code,
        Dimensions = dimensions,
        Chain = chain,
        Count = count,
        AttributedThroughSeq = attributedThroughSeq,
        AsOf = asOf,
    };

    private static long Decimal(string s, int start, int length)
    {
        long value = 0;
        for (int i = start; i < start + length; i++) value = value * 10 + (s[i] - '0');
        return value;
    }

    private static long DaysFromCivil(long year, long month, long day)
    {
        year -= month <= 2 ? 1 : 0;
        long era = year >= 0 ? year / 400 : (year - 399) / 400;
        long yoe = year - era * 400;
        long mp = month + (month > 2 ? -3 : 9);
        long doy = (153 * mp + 2) / 5 + day - 1;
        long doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
        return era * 146097 + doe - 719468;
    }

    /// UTC nanoseconds. Leap-second checkpoints return null, matching the TypeScript Date.parse gate.
    private static BigInteger? ParseInstant(string value)
    {
        if (!Schema.Rfc3339Instant(value)) return null;
        long second = Decimal(value, 17, 2);
        if (second > 59) return null;
        long year = Decimal(value, 0, 4), month = Decimal(value, 5, 2), day = Decimal(value, 8, 2);
        long hour = Decimal(value, 11, 2), minute = Decimal(value, 14, 2);
        int index = 19;
        BigInteger nanos = BigInteger.Zero;
        if (value[index] == '.')
        {
            int start = ++index;
            while (index < value.Length && value[index] >= '0' && value[index] <= '9') index++;
            long fraction = Decimal(value, start, index - start);
            nanos = fraction * BigInteger.Pow(10, 9 - (index - start));
        }
        long offsetSeconds = 0;
        if (value[index] != 'Z' && value[index] != 'z')
        {
            long sign = value[index] == '+' ? 1 : -1;
            offsetSeconds = sign * (Decimal(value, index + 1, 2) * 3600 + Decimal(value, index + 4, 2) * 60);
        }
        long seconds = DaysFromCivil(year, month, day) * 86400 + hour * 3600 + minute * 60 + second - offsetSeconds;
        return new BigInteger(seconds) * 1_000_000_000 + nanos;
    }

    private static HistoricalTrust? ParseKeyring(JVal document)
    {
        if (document is not JObj root) return null;
        bool looksLifecycle = root.Get("spec") is JStr spec && spec.Value == LifecycleSpec
            || root.Has("keys") && root.Get("keys") is not JStr;
        if (!looksLifecycle)
        {
            var flat = new JObj();
            foreach (string kid in root.Keys)
            {
                if (kid.Length == 0 || root.Get(kid) is not JStr publicKey || publicKey.Value.Length == 0) return null;
                flat.Add(kid, new JStr(publicKey.Value));
            }
            return new HistoricalTrust
            {
                Keyring = flat,
                Retired = new Dictionary<string, bool>(StringComparer.Ordinal),
                ValidFrom = new Dictionary<string, BigInteger?>(StringComparer.Ordinal),
                RetiredAt = new Dictionary<string, BigInteger?>(StringComparer.Ordinal),
                Lifecycle = false,
            };
        }
        if (root.Keys.Count != 2 || root.Get("spec") is not JStr lifecycleSpec || lifecycleSpec.Value != LifecycleSpec
            || root.Get("keys") is not JObj entries || entries.Keys.Count == 0) return null;
        var result = new HistoricalTrust
        {
            Keyring = new JObj(),
            Retired = new Dictionary<string, bool>(StringComparer.Ordinal),
            ValidFrom = new Dictionary<string, BigInteger?>(StringComparer.Ordinal),
            RetiredAt = new Dictionary<string, BigInteger?>(StringComparer.Ordinal),
            Lifecycle = true,
        };
        foreach (string kid in entries.Keys)
        {
            if (kid.Length == 0 || entries.Get(kid) is not JObj entry || (entry.Keys.Count != 2 && entry.Keys.Count != 3)
                || !entry.Has("publicKey") || !entry.Has("retiredAt")
                || (entry.Keys.Count == 3 && !entry.Has("validFrom"))
                || entry.Get("publicKey") is not JStr publicKey || publicKey.Value.Length == 0) return null;
            result.Keyring.Add(kid, new JStr(publicKey.Value));
            if (!entry.Has("validFrom") || entry.Get("validFrom") is JNull)
            {
                result.ValidFrom[kid] = null;
            }
            else if (entry.Get("validFrom") is JStr validFrom && ParseInstant(validFrom.Value) is BigInteger activation)
            {
                result.ValidFrom[kid] = activation;
            }
            else return null;
            if (entry.Get("retiredAt") is JNull)
            {
                result.RetiredAt[kid] = null;
            }
            else if (entry.Get("retiredAt") is JStr retiredAt && ParseInstant(retiredAt.Value) is BigInteger instant)
            {
                result.Retired[kid] = true;
                result.RetiredAt[kid] = instant;
            }
            else return null;
            if (result.ValidFrom[kid].HasValue && result.RetiredAt[kid].HasValue
                && result.ValidFrom[kid]!.Value >= result.RetiredAt[kid]!.Value) return null;
        }
        return result;
    }

    public static HistoricalResult Verify(
        JVal receiptsValue,
        JVal receiptRoot,
        JVal? checkpointValue,
        JVal? checkpointRoot,
        JVal? identity)
    {
        string witness = checkpointValue is null ? "NOT_PROVIDED" : "PROVIDED";
        string availability = checkpointValue is null ? "NOT_PROVIDED" : "AVAILABLE";
        JArr? receiptArray = receiptsValue as JArr;
        int count = receiptArray?.Items.Count ?? 0;
        string? chain = receiptArray?.Items.FirstOrDefault() is JObj first
            && first.Get("scope") is JObj scope && scope.Get("chain") is JStr chainString ? chainString.Value : null;
        HistoricalTrust? receiptTrust = ParseKeyring(receiptRoot);
        if (receiptTrust is null)
            return Result("INVALID", "RECEIPT_ROOT_INVALID", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", "NOT_PROVIDED", witness, availability), null, 0);
        string retirement = receiptTrust.Lifecycle ? "PROVIDED" : "NOT_PROVIDED";

        // CONTROL G2-RETIREMENT-INTEGRITY-CSHARP: retained keys reach only the unchanged chain
        // verifier without a checkpoint. Current-use authorization never receives this projection.
        VerifyResult receiptResult = Verifier.VerifyChain(receiptsValue, receiptTrust.Keyring, identity, null);
        if (receiptResult.Status != VerifyStatus.Valid)
        {
            if (receiptResult.Status == VerifyStatus.Tampered)
                return Result("INVALID", "RECEIPT_INTEGRITY_FAILURE", Dimensions("BROKEN", "UNANSWERED", "UNATTRIBUTABLE", retirement, witness, availability), chain, count);
            if (receiptResult.Status == VerifyStatus.Untrusted)
                return Result("INVALID", "RECEIPT_IDENTITY_UNTRUSTED", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", retirement, witness, availability), chain, count);
            return Result("INVALID", "RECEIPT_MALFORMED", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", retirement, witness, availability), chain, count);
        }

        var receipts = ((JArr)receiptsValue).Items.Cast<JObj>().ToList();
        var bySeq = receipts.ToDictionary(
            receipt => ((JInt)((JObj)receipt.Get("chain")!).Get("seq")!).Value,
            receipt => receipt);
        JObj head = bySeq[receipts.Count - 1];
        if (checkpointValue is null)
            return Result("UNVERIFIED", "NO_WITNESS", Dimensions("INTACT", "UNANSWERED", "UNATTRIBUTABLE", retirement, "NOT_PROVIDED", "NOT_PROVIDED"), chain, count);
        if (checkpointRoot is null)
            return Result("UNVERIFIED", "WITNESS_ROOT_NOT_PROVIDED", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        HistoricalTrust? witnessTrust = ParseKeyring(checkpointRoot);
        if (witnessTrust is null)
            return Result("UNVERIFIED", "WITNESS_ROOT_INVALID", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        if (!Verifier.CheckpointShapeOk(checkpointValue))
            return Result("INVALID", "WITNESS_MALFORMED", Dimensions("BROKEN", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        Verifier.CpVerdict cpVerdict = Verifier.VerifyCheckpoint(checkpointValue, witnessTrust.Keyring);
        if (cpVerdict == Verifier.CpVerdict.Unverified)
            return Result("UNVERIFIED", "WITNESS_KEY_NOT_TRUSTED", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        if (cpVerdict != Verifier.CpVerdict.Ok)
            return Result("INVALID", "WITNESS_INTEGRITY_FAILURE", Dimensions("BROKEN", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);

        var checkpoint = (JObj)checkpointValue;
        var checkpointSig = (JObj)checkpoint.Get("sig")!;
        string witnessKid = ((JStr)checkpointSig.Get("kid")!).Value;
        string witnessPublic = ((JStr)witnessTrust.Keyring.Get(witnessKid)!).Value;
        byte[] witnessMaterial;
        try { witnessMaterial = Crypto.SpkiToRaw(witnessPublic); }
        catch { return Result("UNVERIFIED", "WITNESS_ROOT_INVALID", Dimensions("UNANSWERED", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count); }

        // CONTROL G2-WITNESS-KEY-SEPARATION-CSHARP: compare decoded key material, not kid labels.
        foreach (JObj receipt in receipts)
        {
            string receiptKid = ((JStr)((JObj)receipt.Get("sig")!).Get("kid")!).Value;
            string receiptPublic = ((JStr)receiptTrust.Keyring.Get(receiptKid)!).Value;
            if (witnessKid == receiptKid || witnessMaterial.AsSpan().SequenceEqual(Crypto.SpkiToRaw(receiptPublic)))
                return Result("UNVERIFIED", "WITNESS_KEY_NOT_SEPARATE", Dimensions("INTACT", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        }
        if (witnessTrust.Retired.ContainsKey(witnessKid))
            return Result("UNVERIFIED", "WITNESS_KEY_RETIRED", Dimensions("INTACT", "UNANSWERED", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);

        string checkpointChain = ((JStr)checkpoint.Get("chain")!).Value;
        if (checkpointChain != chain)
            return Result("CONFLICT", "CHECKPOINT_CONFLICT", Dimensions("INTACT", "CONFLICT", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        long checkpointSeq = ((JInt)checkpoint.Get("highestSeq")!).Value;
        long headSeq = ((JInt)((JObj)head.Get("chain")!).Get("seq")!).Value;
        if (checkpointSeq > headSeq)
            return Result("CONFLICT", "CHECKPOINT_AHEAD", Dimensions("INTACT", "CONFLICT", "UNATTRIBUTABLE", retirement, "PROVIDED", "MISSING_RELATIVE_TO_CHECKPOINT"), chain, count);
        string checkpointHash = ((JStr)checkpoint.Get("headHash")!).Value;
        if (!bySeq.TryGetValue(checkpointSeq, out JObj? checkpointed)
            || ((JStr)((JObj)checkpointed.Get("chain")!).Get("hash")!).Value != checkpointHash)
            return Result("CONFLICT", "CHECKPOINT_CONFLICT", Dimensions("INTACT", "CONFLICT", "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);

        string completeness = checkpointSeq == headSeq ? "HEAD_ANCHORED" : "PREFIX_ANCHORED";
        string checkpointTimeText = ((JStr)checkpoint.Get("ts")!).Value;
        BigInteger? checkpointTime = ParseInstant(checkpointTimeText);
        bool checkpointBeforeActivation = witnessTrust.ValidFrom.TryGetValue(witnessKid, out BigInteger? witnessValidFrom)
            && witnessValidFrom.HasValue
            && (!checkpointTime.HasValue || checkpointTime.Value < witnessValidFrom.Value);
        bool checkpointAfterRetirement = !checkpointTime.HasValue;
        for (long seq = 0; seq <= checkpointSeq; seq++)
        {
            string receiptKid = ((JStr)((JObj)bySeq[seq].Get("sig")!).Get("kid")!).Value;
            if (receiptTrust.ValidFrom.TryGetValue(receiptKid, out BigInteger? validFrom) && validFrom.HasValue
                && (!checkpointTime.HasValue || checkpointTime.Value < validFrom.Value)) checkpointBeforeActivation = true;
            if (receiptTrust.RetiredAt.TryGetValue(receiptKid, out BigInteger? retiredAt) && retiredAt.HasValue
                && (!checkpointTime.HasValue || checkpointTime.Value >= retiredAt.Value)) checkpointAfterRetirement = true;
        }
        if (checkpointBeforeActivation)
            return Result("UNVERIFIED", "CHECKPOINT_BEFORE_ACTIVATION", Dimensions("INTACT", completeness, "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);
        if (checkpointAfterRetirement)
            return Result("UNVERIFIED", "CHECKPOINT_AFTER_RETIREMENT", Dimensions("INTACT", completeness, "UNATTRIBUTABLE", retirement, "PROVIDED", "AVAILABLE"), chain, count);

        return Result(
            completeness == "HEAD_ANCHORED" ? "VERIFIED" : "PARTIAL",
            completeness,
            Dimensions("INTACT", completeness, "ATTRIBUTABLE_AS_OF", retirement, "PROVIDED", "AVAILABLE"),
            chain,
            count,
            checkpointSeq,
            checkpointTimeText);
    }
}

File: impl-rust/src/historical.rs
//! Versioned historical verification side API. It authenticates retained receipt signatures and a
//! separately rooted checkpoint without changing the current-use `verify_chain` contract.

use crate::json::Json;
use crate::keys::spki_to_raw;
use crate::schema::is_rfc3339_instant;
use crate::verify::{checkpoint_shape_ok, verify_chain, verify_checkpoint, Status};
use serde_json::{json, Value};
use std::collections::HashMap;

const HISTORICAL_SPEC: &str = "noa.historical-verification/0.1";
const LIFECYCLE_SPEC: &str = "noa.signing-key-lifecycle/0.1";

struct HistoricalTrust {
    keyring: Json,
    retired: HashMap<String, bool>,
    valid_from: HashMap<String, Option<i128>>,
    retired_at: HashMap<String, Option<i128>>,
    lifecycle: bool,
}

fn dimensions(
    integrity: &str,
    completeness: &str,
    attribution: &str,
    retirement: &str,
    witness: &str,
    availability: &str,
) -> Value {
    json!({
        "integrity": integrity,
        "completeness": completeness,
        "attribution": attribution,
        "organizationalIndependence": "UNVERIFIED",
        "evidence": {
            "retirement": retirement,
            "witness": witness,
            "availability": availability
        }
    })
}

fn result(
    classification: &str,
    code: &str,
    dimensions: Value,
    chain: Option<&str>,
    count: usize,
    attributed_through_seq: Option<i64>,
    as_of: Option<&str>,
) -> Value {
    json!({
        "spec": HISTORICAL_SPEC,
        "policy": {"verifierVersion": HISTORICAL_SPEC, "purpose": "historical-audit"},
        "classification": classification,
        "code": code,
        "dimensions": dimensions,
        "chain": chain,
        "count": count,
        "attributedThroughSeq": attributed_through_seq,
        "asOf": as_of
    })
}

fn days_from_civil(mut year: i64, month: i64, day: i64) -> i64 {
    year -= if month <= 2 { 1 } else { 0 };
    let era = if year >= 0 {
        year / 400
    } else {
        (year - 399) / 400
    };
    let yoe = year - era * 400;
    let mp = month + if month > 2 { -3 } else { 9 };
    let doy = (153 * mp + 2) / 5 + day - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

fn decimal(bytes: &[u8]) -> Option<i64> {
    let mut out = 0i64;
    for b in bytes {
        if !b.is_ascii_digit() {
            return None;
        }
        out = out * 10 + i64::from(*b - b'0');
    }
    Some(out)
}

/// UTC nanoseconds. Leap-second checkpoints return None, matching the TypeScript Date.parse gate.
fn parse_instant(value: &str) -> Option<i128> {
    if !is_rfc3339_instant(value) {
        return None;
    }
    let b = value.as_bytes();
    let year = decimal(&b[0..4])?;
    let month = decimal(&b[5..7])?;
    let day = decimal(&b[8..10])?;
    let hour = decimal(&b[11..13])?;
    let minute = decimal(&b[14..16])?;
    let second = decimal(&b[17..19])?;
    if second > 59 {
        return None;
    }
    let mut i = 19usize;
    let mut nanos = 0i128;
    if b.get(i) == Some(&b'.') {
        i += 1;
        let start = i;
        while i < b.len() && b[i].is_ascii_digit() {
            i += 1;
        }
        let fraction = decimal(&b[start..i])? as i128;
        nanos = fraction * 10i128.pow((9 - (i - start)) as u32);
    }
    let offset_seconds = if b.get(i) == Some(&b'Z') || b.get(i) == Some(&b'z') {
        0i64
    } else {
        let sign = if b.get(i) == Some(&b'+') { 1i64 } else { -1i64 };
        sign * (decimal(&b[i + 1..i + 3])? * 3600 + decimal(&b[i + 4..i + 6])? * 60)
    };
    let seconds = days_from_civil(year, month, day) * 86400 + hour * 3600 + minute * 60 + second
        - offset_seconds;
    Some(i128::from(seconds) * 1_000_000_000 + nanos)
}

fn parse_keyring(document: &Json) -> Option<HistoricalTrust> {
    let object = document.as_object()?;
    let looks_lifecycle = document.get("spec").and_then(Json::as_str) == Some(LIFECYCLE_SPEC)
        || matches!(document.get("keys"), Some(v) if !matches!(v, Json::Str(_)));
    if !looks_lifecycle {
        let mut flat = Vec::with_capacity(object.len());
        for (kid, public_key) in object {
            let value = public_key.as_str()?;
            if kid.is_empty() || value.is_empty() {
                return None;
            }
            flat.push((kid.clone(), Json::Str(value.to_string())));
        }
        return Some(HistoricalTrust {
            keyring: Json::Object(flat),
            retired: HashMap::new(),
            valid_from: HashMap::new(),
            retired_at: HashMap::new(),
            lifecycle: false,
        });
    }
    if object.len() != 2 || document.get("spec").and_then(Json::as_str) != Some(LIFECYCLE_SPEC) {
        return None;
    }
    let entries = document.get("keys")?.as_object()?;
    if entries.is_empty() {
        return None;
    }
    let mut flat = Vec::with_capacity(entries.len());
    let mut retired = HashMap::new();
    let mut valid_from = HashMap::new();
    let mut retired_at = HashMap::new();
    for (kid, entry) in entries {
        let fields = entry.as_object()?;
        if kid.is_empty()
            || (fields.len() != 2 && fields.len() != 3)
            || entry.get("publicKey").is_none()
            || entry.get("retiredAt").is_none()
            || (fields.len() == 3 && entry.get("validFrom").is_none())
        {
            return None;
        }
        let public_key = entry.get("publicKey")?.as_str()?;
        if public_key.is_empty() {
            return None;
        }
        flat.push((kid.clone(), Json::Str(public_key.to_string())));
        let activation = match entry.get("validFrom") {
            None | Some(Json::Null) => None,
            Some(Json::Str(value)) => Some(parse_instant(value)?),
            _ => return None,
        };
        let retirement = match entry.get("retiredAt")? {
            Json::Null => None,
            Json::Str(value) => {
                let instant = parse_instant(value)?;
                retired.insert(kid.clone(), true);
                Some(instant)
            }
            _ => return None,
        };
        if matches!((activation, retirement), (Some(start), Some(end)) if start >= end) {
            return None;
        }
        valid_from.insert(kid.clone(), activation);
        retired_at.insert(kid.clone(), retirement);
    }
    Some(HistoricalTrust {
        keyring: Json::Object(flat),
        retired,
        valid_from,
        retired_at,
        lifecycle: true,
    })
}

pub fn verify_historical_chain(
    receipts: &Json,
    receipt_root: &Json,
    checkpoint: Option<&Json>,
    checkpoint_root: Option<&Json>,
    identity: Option<&Json>,
) -> Value {
    let witness = if checkpoint.is_some() {
        "PROVIDED"
    } else {
        "NOT_PROVIDED"
    };
    let availability = if checkpoint.is_some() {
        "AVAILABLE"
    } else {
        "NOT_PROVIDED"
    };
    let (arr, count, chain) = match receipts {
        Json::Array(values) => (
            Some(values),
            values.len(),
            values
                .first()
                .and_then(|r| r.get("scope"))
                .and_then(|s| s.get("chain"))
                .and_then(Json::as_str),
        ),
        _ => (None, 0, None),
    };
    let receipt_trust = match parse_keyring(receipt_root) {
        Some(value) => value,
        None => {
            return result(
                "INVALID",
                "RECEIPT_ROOT_INVALID",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    "NOT_PROVIDED",
                    witness,
                    availability,
                ),
                None,
                0,
                None,
                None,
            )
        }
    };
    let retirement = if receipt_trust.lifecycle {
        "PROVIDED"
    } else {
        "NOT_PROVIDED"
    };

    // CONTROL G2-RETIREMENT-INTEGRITY-RUST: only this historical side API projects retained keys
    // into the unchanged current verifier, and it never forwards a checkpoint through receipt trust.
    let (status, _) = verify_chain(receipts, Some(&receipt_trust.keyring), identity, None);
    if status != Status::Valid {
        return match status {
            Status::Tampered => result(
                "INVALID",
                "RECEIPT_INTEGRITY_FAILURE",
                dimensions(
                    "BROKEN",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    witness,
                    availability,
                ),
                chain,
                count,
                None,
                None,
            ),
            Status::Untrusted => result(
                "INVALID",
                "RECEIPT_IDENTITY_UNTRUSTED",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    witness,
                    availability,
                ),
                chain,
                count,
                None,
                None,
            ),
            _ => result(
                "INVALID",
                "RECEIPT_MALFORMED",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    witness,
                    availability,
                ),
                chain,
                count,
                None,
                None,
            ),
        };
    }
    let arr = arr.unwrap();
    let mut by_seq: HashMap<i64, &Json> = HashMap::new();
    for receipt in arr {
        by_seq.insert(
            receipt
                .get("chain")
                .and_then(|c| c.get("seq"))
                .and_then(Json::as_int)
                .unwrap(),
            receipt,
        );
    }
    let head = by_seq.get(&((count - 1) as i64)).unwrap();
    let checkpoint = match checkpoint {
        Some(value) => value,
        None => {
            return result(
                "UNVERIFIED",
                "NO_WITNESS",
                dimensions(
                    "INTACT",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "NOT_PROVIDED",
                    "NOT_PROVIDED",
                ),
                chain,
                count,
                None,
                None,
            )
        }
    };
    let checkpoint_root = match checkpoint_root {
        Some(value) => value,
        None => {
            return result(
                "UNVERIFIED",
                "WITNESS_ROOT_NOT_PROVIDED",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "PROVIDED",
                    "AVAILABLE",
                ),
                chain,
                count,
                None,
                None,
            )
        }
    };
    let witness_trust = match parse_keyring(checkpoint_root) {
        Some(value) => value,
        None => {
            return result(
                "UNVERIFIED",
                "WITNESS_ROOT_INVALID",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "PROVIDED",
                    "AVAILABLE",
                ),
                chain,
                count,
                None,
                None,
            )
        }
    };
    if !checkpoint_shape_ok(checkpoint) {
        return result(
            "INVALID",
            "WITNESS_MALFORMED",
            dimensions(
                "BROKEN",
                "UNANSWERED",
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "AVAILABLE",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    match verify_checkpoint(checkpoint, Some(&witness_trust.keyring)) {
        "unverified" => {
            return result(
                "UNVERIFIED",
                "WITNESS_KEY_NOT_TRUSTED",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "PROVIDED",
                    "AVAILABLE",
                ),
                chain,
                count,
                None,
                None,
            )
        }
        "ok" => {}
        _ => {
            return result(
                "INVALID",
                "WITNESS_INTEGRITY_FAILURE",
                dimensions(
                    "BROKEN",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "PROVIDED",
                    "AVAILABLE",
                ),
                chain,
                count,
                None,
                None,
            )
        }
    }

    let witness_kid = checkpoint
        .get("sig")
        .and_then(|s| s.get("kid"))
        .and_then(Json::as_str)
        .unwrap();
    let witness_public = witness_trust
        .keyring
        .get(witness_kid)
        .and_then(Json::as_str)
        .unwrap();
    let witness_material = match spki_to_raw(witness_public) {
        Ok(value) => value,
        Err(_) => {
            return result(
                "UNVERIFIED",
                "WITNESS_ROOT_INVALID",
                dimensions(
                    "UNANSWERED",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "PROVIDED",
                    "AVAILABLE",
                ),
                chain,
                count,
                None,
                None,
            )
        }
    };
    // CONTROL G2-WITNESS-KEY-SEPARATION-RUST: compare canonical decoded SPKI key bytes, not labels.
    for receipt in arr {
        let receipt_kid = receipt
            .get("sig")
            .and_then(|s| s.get("kid"))
            .and_then(Json::as_str)
            .unwrap();
        let receipt_public = receipt_trust
            .keyring
            .get(receipt_kid)
            .and_then(Json::as_str)
            .unwrap();
        if witness_kid == receipt_kid
            || spki_to_raw(receipt_public).ok().as_ref() == Some(&witness_material)
        {
            return result(
                "UNVERIFIED",
                "WITNESS_KEY_NOT_SEPARATE",
                dimensions(
                    "INTACT",
                    "UNANSWERED",
                    "UNATTRIBUTABLE",
                    retirement,
                    "PROVIDED",
                    "AVAILABLE",
                ),
                chain,
                count,
                None,
                None,
            );
        }
    }
    if witness_trust
        .retired
        .get(witness_kid)
        .copied()
        .unwrap_or(false)
    {
        return result(
            "UNVERIFIED",
            "WITNESS_KEY_RETIRED",
            dimensions(
                "INTACT",
                "UNANSWERED",
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "AVAILABLE",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    if checkpoint.get("chain").and_then(Json::as_str) != chain {
        return result(
            "CONFLICT",
            "CHECKPOINT_CONFLICT",
            dimensions(
                "INTACT",
                "CONFLICT",
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "AVAILABLE",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    let cp_seq = checkpoint.get("highestSeq").and_then(Json::as_int).unwrap();
    let head_seq = head
        .get("chain")
        .and_then(|c| c.get("seq"))
        .and_then(Json::as_int)
        .unwrap();
    if cp_seq > head_seq {
        return result(
            "CONFLICT",
            "CHECKPOINT_AHEAD",
            dimensions(
                "INTACT",
                "CONFLICT",
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "MISSING_RELATIVE_TO_CHECKPOINT",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    let checkpointed = by_seq.get(&cp_seq);
    let hash_matches = checkpointed
        .and_then(|r| r.get("chain"))
        .and_then(|c| c.get("hash"))
        .and_then(Json::as_str)
        == checkpoint.get("headHash").and_then(Json::as_str);
    if checkpointed.is_none() || !hash_matches {
        return result(
            "CONFLICT",
            "CHECKPOINT_CONFLICT",
            dimensions(
                "INTACT",
                "CONFLICT",
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "AVAILABLE",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    let completeness = if cp_seq == head_seq {
        "HEAD_ANCHORED"
    } else {
        "PREFIX_ANCHORED"
    };
    let checkpoint_time = checkpoint
        .get("ts")
        .and_then(Json::as_str)
        .and_then(parse_instant);
    let mut checkpoint_before_activation =
        if let Some(Some(valid_from)) = witness_trust.valid_from.get(witness_kid) {
            checkpoint_time
                .map(|value| value < *valid_from)
                .unwrap_or(true)
        } else {
            false
        };
    let mut checkpoint_after_retirement = checkpoint_time.is_none();
    for seq in 0..=cp_seq {
        let receipt_kid = by_seq[&seq]
            .get("sig")
            .and_then(|s| s.get("kid"))
            .and_then(Json::as_str)
            .unwrap();
        if let Some(Some(valid_from)) = receipt_trust.valid_from.get(receipt_kid) {
            if checkpoint_time
                .map(|value| value < *valid_from)
                .unwrap_or(true)
            {
                checkpoint_before_activation = true;
            }
        }
        if let Some(Some(retired_at)) = receipt_trust.retired_at.get(receipt_kid) {
            if checkpoint_time
                .map(|value| value >= *retired_at)
                .unwrap_or(true)
            {
                checkpoint_after_retirement = true;
            }
        }
    }
    if checkpoint_before_activation {
        return result(
            "UNVERIFIED",
            "CHECKPOINT_BEFORE_ACTIVATION",
            dimensions(
                "INTACT",
                completeness,
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "AVAILABLE",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    if checkpoint_after_retirement {
        return result(
            "UNVERIFIED",
            "CHECKPOINT_AFTER_RETIREMENT",
            dimensions(
                "INTACT",
                completeness,
                "UNATTRIBUTABLE",
                retirement,
                "PROVIDED",
                "AVAILABLE",
            ),
            chain,
            count,
            None,
            None,
        );
    }
    let classification = if completeness == "HEAD_ANCHORED" {
        "VERIFIED"
    } else {
        "PARTIAL"
    };
    let as_of = checkpoint.get("ts").and_then(Json::as_str);
    result(
        classification,
        completeness,
        dimensions(
            "INTACT",
            completeness,
            "ATTRIBUTABLE_AS_OF",
            retirement,
            "PROVIDED",
            "AVAILABLE",
        ),
        chain,
        count,
        Some(cp_seq),
        as_of,
    )
}
